File: haskell/glue/src/Glue/Eval/Exception.hs
module Glue.Eval.Exception where

import Data.Text (Text)
import Data.Text qualified as T
import Glue.IR (IR (..))

data RuntimeException m
    = RuntimeException Text (Maybe (IR m))
    deriving (Show, Eq)

pretty :: RuntimeException m -> Text
pretty = \case
    RuntimeException sym msg -> "Runtime Exception: " <> sym <> ". " <> T.pack (show msg)

unboundVariable :: Text -> RuntimeException m
unboundVariable = RuntimeException "unbound-variable" . Just . String

canNotSetUnboundVariable :: Text -> RuntimeException m
canNotSetUnboundVariable = RuntimeException "cannot-set-unbound-variable" . Just . String

notCallableObject :: RuntimeException m
notCallableObject = RuntimeException "not-callable-object" Nothing

expectedValue :: RuntimeException m
expectedValue = RuntimeException "expected-value" Nothing

expectedListOfSymbols :: RuntimeException m
expectedListOfSymbols = RuntimeException "expected-list-of-symbols" Nothing

wrongNumberOfArguments :: RuntimeException m
wrongNumberOfArguments = RuntimeException "wrong-number-of-arguments" Nothing

wrongArgumentType :: [Text] -> RuntimeException m
wrongArgumentType expected = RuntimeException "wrong-argument-type" . Just $ List (String <$> expected)

divByZero :: RuntimeException m
divByZero = RuntimeException "div-by-zero" Nothing

propertyNotFound :: Text -> RuntimeException m
propertyNotFound = RuntimeException "property-not-found" . Just . String

notAnObject :: IR m -> RuntimeException m
notAnObject = RuntimeException "not-an-object" . Just

moduleNotFound :: Text -> RuntimeException m
moduleNotFound = RuntimeException "module-not-found" . Just . String

undefinedExport :: Text -> RuntimeException m
undefinedExport = RuntimeException "undefined-export" . Just . String

runtimeException :: Text -> IR m -> RuntimeException m
runtimeException symbol = RuntimeException symbol . Just

File: dart/glue/lib/src/eval/exception.dart
import 'package:glue/src/ir.dart';

/// Runtime exceptions for Glue evaluation
/// Mirrors Haskell Glue.Eval.Exception exactly
class RuntimeException {
  final String symbol;
  final Ir? value;

  const RuntimeException(this.symbol, this.value);

  String pretty() => 'Runtime Exception: $symbol. ${value ?? 'No details'}';

  @override
  String toString() => pretty();

  @override
  bool operator ==(Object other) =>
      other is RuntimeException &&
      other.symbol == symbol &&
      other.value == value;

  @override
  int get hashCode => Object.hash(symbol, value);
}

/// Factory functions for specific runtime exceptions
/// Mirrors Haskell constructor functions exactly

RuntimeException unboundVariable(String name) =>
    RuntimeException('unbound-variable', IrString(name));

RuntimeException canNotSetUnboundVariable(String name) =>
    RuntimeException('cannot-set-unbound-variable', IrString(name));

RuntimeException notCallableObject() =>
    RuntimeException('not-callable-object', null);

RuntimeException expectedValue() => RuntimeException('expected-value', null);

RuntimeException expectedListOfSymbols() =>
    RuntimeException('expected-list-of-symbols', null);

RuntimeException wrongNumberOfArguments() =>
    RuntimeException('wrong-number-of-arguments', null);

RuntimeException wrongArgumentType(List<String> expected) => RuntimeException(
  'wrong-argument-type',
  IrList(expected.map(IrString.new).toList()),
);

RuntimeException divByZero() => RuntimeException('div-by-zero', null);

RuntimeException propertyNotFound(String property) =>
    RuntimeException('property-not-found', IrString(property));

RuntimeException notAnObject(Ir ir) => RuntimeException('not-an-object', ir);

RuntimeException moduleNotFound(String module) =>
    RuntimeException('module-not-found', IrString(module));

RuntimeException runtimeException(String symbol, Ir value) =>
    RuntimeException(symbol, value);
